File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            while (left < right && !checkAlphaNumeric(s[left])) left++;
            while (left < right && !checkAlphaNumeric(s[right])) right--;

            char l = tolower(s[left]);
            char r = tolower(s[right]);

            if (l != r) return false;

            left++;
            right--;
        }

        return true;
    }

    bool checkAlphaNumeric(char a) {
        return (a >= 'A' && a <= 'Z') || 
               (a >= 'a' && a <= 'z') || 
               (a >= '0' && a <= '9');
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;

        int [] left = new int[n];
        int [] right = new int[n];

        Stack<Integer>st = new Stack<>();

        for(int i = 0; i<n; i++){
            while(!st.empty() && heights[st.peek()] >= heights[i]) st.pop();
            left[i] = st.empty()?-1:st.peek();
            st.push(i);
        }

        st.clear();

        for(int i = n-1; i>=0; i--){
            while(!st.empty() && heights[st.peek()] >= heights[i]) st.pop();
            right[i] = st.empty()?n:st.peek();
            st.push(i);
        }

        int maxArea = 0;

        for(int i=0;i<n;i++){
            int width = right[i] - left[i] - 1;
            maxArea = Math.max(maxArea, width * heights[i]);
        }

        return maxArea;
    }
}
